gsd/gsd.c

```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
+*/
+static int gsd_index_buffer_free(struct gsd_index_buffer* buf)
+{
+    if (buf == NULL || buf->data == NULL)
     {
-        handle->cur_frame = 0;
+        return GSD_ERROR_INVALID_ARGUMENT;
+    }
+
+#if GSD_USE_MMAP
+    if (buf->mapped_data)
+    {
+        int retval = munmap(buf->mapped_data, buf->mapped_len);
+
+        if (retval != 0)
+        {
+            return GSD_ERROR_IO;
+        }
     }
     else
+#endif
     {
-        handle->cur_frame = handle->index[handle->index_num_entries - 1].frame + 1;
+        free(buf->data);
     }
 
-    // at this point, all valid index entries have been written to disk
-    handle->index_written_entries = handle->index_num_entries;
+    gsd_util_zero_memory(buf, sizeof(struct gsd_index_buffer));
+    return GSD_SUCCESS;
+}
+
+/** @internal
+    @brief Add a new index entry and provide a pointer to it.
+
+    @param buf Buffer to add too.
+    @param entry [out] Pointer to set to the new entry.
 
-    if (handle->open_flags == GSD_OPEN_APPEND)
+    Double the size of the reserved space as needed to hold the new entry. Does not accept mapped
+    indices.
+
+    @returns GSD_SUCCESS on success, GSD_* error codes on error.
+*/
+inline static int gsd_index_buffer_add(struct gsd_index_buffer* buf, struct gsd_index_entry** entry)
+{
+    if (buf == NULL || buf->mapped_data || entry == NULL || buf->reserved == 0)
     {
-#if GSD_USE_MMAP
-        // in append mode, we need to tear down the temporary mapping and allocate a temporary
-        // buffer to hold indices for a single frame
-        size_t page_size = getpagesize();
-        size_t index_size = sizeof(struct gsd_index_entry) * handle->header.index_allocated_entries;
-        size_t offset = (handle->header.index_location / page_size) * page_size;
-        int retval
-            = munmap(handle->mapped_data, index_size + (handle->header.index_location - offset));
-        handle->index = NULL;
+        return GSD_ERROR_INVALID_ARGUMENT;
+    }
 
-        if (retval != 0)
+    if (buf->size == buf->reserved)
+    {
+        // grow the array
+        size_t new_reserved = buf->reserved * 2;
+        buf->data = realloc(buf->data, sizeof(struct gsd_index_entry) * new_reserved);
+        if (buf->data == NULL)
+        {
+            return GSD_ERROR_MEMORY_ALLOCATION_FAILED;
+        }
+
+        // zero the new memory
+        gsd_util_zero_memory(buf->data + buf->reserved,
+                             sizeof(struct gsd_index_entry) * (new_reserved - buf->reserved));
+        buf->reserved = new_reserved;
+    }
+
+    size_t insert_pos = buf->size;
+    buf->size++;
+    *entry = buf->data + insert_pos;
+
+    return GSD_SUCCESS;
+}
+
+inline static int gsd_cmp_index_entry(const void* p1, const void* p2)
+{
+    struct gsd_index_entry* a = (struct gsd_index_entry*)p1;
+    struct gsd_index_entry* b = (struct gsd_index_entry*)p2;
+    int result = 0;
+
+    if (a->frame < b->frame)
+    {
+        result = -1;
```

Comment:
I'd suggest just returning inside each conditional (or changing everything to be if/else clauses) to reduce the number of conditional evaluations) since this function is probably called a lot at a very deep part of the code during any heap operations.

Reply:
clang-tidy complains about if clauses after a return. In either case the optimizer is smart and optimizes away the result value. When I looked at this in `perf` output, I saw that the `jmp` instructions for the loop calling `gsd_cmp_index_entry` are placed directly inside the conditionals.

Reply:
OK, I'm always a little wary of trusting compiler optimizations but that's mostly because I haven't looked at the generated assembly enough to see how well they do. I trust your judgement there.